rust/src/sdk/api.rs

```diff
@@ -1,171 +1,243 @@
 /// of its `input` across the FFI-boundary.
 pub unsafe extern "C" fn new_client(
     period: c_ulong,
-    callback: unsafe extern "C" fn(*mut c_void, *const c_void),
-    state: *mut c_void,
-) -> *mut Client {
-    if period == 0 {
-        // TODO: add error handling
-        panic!("polling period must be positive")
+    min_threads: c_ulong,
+    max_threads: c_ulong,
+) -> *mut FFIClient {
+    if period == 0 || min_threads == 0 || max_threads < min_threads || max_threads > 32_768 {
+        return ptr::null_mut() as *mut FFIClient;
     }
-    let client = if let Ok(client) = Client::new(period as u64) {
+    let runtime = if let Ok(runtime) = Builder::new()
+        .threaded_scheduler()
+        .core_threads(min_threads as usize)
+        .max_threads(max_threads as usize)
+        .thread_name("xain-fl-client-runtime-worker")
+        .enable_all()
+        .build()
+    {
+        runtime
+    } else {
+        return ptr::null_mut() as *mut FFIClient;
+    };
+    let client = if let Ok(client) = runtime.enter(move || Client::new(period)) {
         client
     } else {
-        // TODO: add error handling
-        panic!("participant initialization failed")
+        return ptr::null_mut() as *mut FFIClient;
     };
+    Box::into_raw(Box::new(FFIClient { runtime, client }))
+}
 
-    // TODO: actually start the client, requires tokio running
-
-    #[repr(C)]
-    struct Input {
-        _round_started: bool,
-        _participant_initialized: bool,
-        _model_cached: bool,
+#[allow(unused_unsafe)]
+#[no_mangle]
+/// Starts the [`Client`] and executes its tasks in an asynchronous [`Runtime`].
+///
+/// Takes a `callback(state, code)` function pointer and a void pointer to the `state` of the
+/// callback. The callback will be triggered when the client must be stopped because of a panic or
+/// error or when the client terminates successfully. See the [errors] section for the error `code`
+/// definitions.
+///
+/// # Errors
+/// Ignores null pointer `client`s and triggers the callback immediately.
+///
+/// Triggers the callback with one of the following error codes:
+/// - `-1`: client didn't start due to null pointer
+/// - `0`: no error (only for clients with finite number of FL rounds)
+/// - `1`: client panicked due to unexpected/unhandled error
+/// - `2`: client stopped due to error [`ParticipantInitErr`]
+/// - `3`: client stopped due to error [`ParticipantErr`]
+/// - `4`: client stopped due to error [`DeserialiseErr`]
+/// - `5`: client stopped due to error [`GeneralErr`]
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+///
+/// If the callback is triggered because of a panicking client, it is the users responsibility
+/// to not access possibly invalid state and to drop the client. If certain parts of the state
```

Comment:
i think part of the ambiguity comes from the `UnwindSafe`ty: i removed the last sentences since it is probably quite impossible the check the `Client`'s invariants from the caller side after a panic (we might do that ourself in the future and then return an error code instead of a panic code). i rephrased the "state" to "client state" to better reflect the meaning.

we may also rename the callback's `state` to sth different to avoid ambiguity.